memory: VariablePointer can panic on malformed variable offset

###  Bug Description

`VariablePointer[T]` can panic instead of returning an error when a variable offset overflows a `uint32` bounds check.

This is behind the `ebpf_unsafe_memory_experiment` build tag. The issue is still worth handling because the exported pointer API already validates mapped memory bounds and should reject malformed offsets instead of crashing the process.

The failing path is:

1. `newVariable` accepts an invalid variable range because `offset + size` wraps.
2. `VariablePointer[T]` calls `memoryPointer[T]`.
3. `checkUnsafeMemory` repeats the wrapped `off + size` check.
4. `memoryPointer` indexes the byte slice at the invalid offset and panics.

Expected behavior:

Malformed variable bounds return an error.

Actual behavior:

The process panics with an out-of-range slice index.

### How to reproduce

On upstream `main`, add this temporary test:

```go
//go:build ebpf_unsafe_memory_experiment

package ebpf

import (
	"math"
	"testing"

	"github.com/go-quicktest/qt"
)

func TestVariablePointerOffsetOverflowRepro(t *testing.T) {
	mm := &Memory{b: make([]byte, 8), heap: true}

	v, err := newVariable("x", math.MaxUint32-3, 4, nil, mm)
	qt.Assert(t, qt.IsNil(err))

	_, err = VariablePointer[uint32](v)
	qt.Assert(t, qt.IsNotNil(err))
}
```

Run:

```sh
go test -tags ebpf_unsafe_memory_experiment . -run TestVariablePointerOffsetOverflowRepro -count=1
```

Observed output:

```text
panic: runtime error: index out of range [4294967292] with length 8

github.com/cilium/ebpf.memoryPointer[...](..., 0xfffffffc)
	.../memory_unsafe.go:342
github.com/cilium/ebpf.VariablePointer[...](...)
	.../variable.go:247
```

### Version information

Reproduced on upstream `main` at `765fc1d`.

### Possible fix

Use overflow-safe bounds checks in `newVariable` and `checkUnsafeMemory`. The existing `Memory.bounds(off, size)` helper already has the right shape, or the checks can be written as `off > len || size > len-off`.

--- a/memory_unsafe.go
+++ b/memory_unsafe.go
@@ -243,7 +243,7 @@ func checkUnsafeMemory[T comparable](mm *Memory, off uint32) error {
 	}
 
 	vs, bs := uint32(size), uint32(len(mm.b))
-	if off+vs > bs {
+	if !mm.bounds(off, vs) {
 		return fmt.Errorf("%d-byte value at offset %d exceeds mmap size of %d bytes", vs, off, bs)
 	}
 

--- a/variable.go
+++ b/variable.go
@@ -125,7 +125,7 @@ type Variable struct {
 
 func newVariable(name string, offset, size uint32, t *btf.Var, mm *Memory) (*Variable, error) {
 	if mm != nil {
-		if offset+size > mm.Size() {
+		if !mm.bounds(offset, size) {
 			return nil, fmt.Errorf("offset %d(+%d) is out of bounds", offset, size)
 		}
 	}

--- a/variable_test.go
+++ b/variable_test.go
@@ -2,6 +2,7 @@ package ebpf
 
 import (
 	"encoding/binary"
+	"math"
 	"runtime"
 	"structs"
 	"sync/atomic"
@@ -295,6 +296,21 @@ func TestVariablePointerError(t *testing.T) {
 
 	_, err = VariablePointer[atomic.Uint32](obj.Atomic)
 	qt.Assert(t, qt.ErrorIs(err, ErrNotSupported))
+}
+
+func TestVariablePointerOffsetOverflow(t *testing.T) {
+	mm := &Memory{b: make([]byte, 8)}
+
+	// An offset whose end wraps around uint32 must be rejected rather than
+	// accepted and later used to index into the backing slice.
+	_, err := newVariable("overflow", math.MaxUint32-3, 4, nil, mm)
+	qt.Assert(t, qt.IsNotNil(err))
+
+	// Even if a Variable carries such an offset, taking a pointer to it returns
+	// an error instead of panicking with an out-of-range index.
+	v := &Variable{name: "overflow", offset: math.MaxUint32 - 3, size: 4, mm: mm}
+	_, err = VariablePointer[uint32](v)
+	qt.Assert(t, qt.IsNotNil(err))
 }
 
 func TestVariablePointerGC(t *testing.T) {

